sql/sql_show.cc

```diff
@@ -8595,6 +8601,231 @@ get_referential_constraints_record(THD *thd, TABLE_LIST *tables,
+  table->field[38]->store(mi->rli.last_error().message,
+                          strlen(mi->rli.last_error().message), cs);
+  prot_store_ids(thd, &mi->ignore_server_ids, table->field[41]);
+  table->field[42]->store((uint32) mi->master_id);
+  table->field[43]->store(mi->ssl_crl , strlen(mi->ssl_crl), cs);
+  table->field[44]->store(mi->ssl_crlpath, strlen(mi->ssl_crlpath), cs);
+  msg= (mi->using_gtid_astext(mi->using_gtid)?
+        mi->using_gtid_astext(mi->using_gtid) : "");
+  table->field[45]->store(msg, strlen(msg), cs);
+  mi->gtid_current_pos.to_string(&str);
+  table->field[46]->store(str.ptr(), str.length(), cs);
+  str.length(0);
+  // Replicate_Ignore_Domain_Ids
+  mi->domain_id_filter.store_ids(thd, table->field[47]);
+  {
+    const char *mode_name= get_type(&slave_parallel_mode_typelib,
+                                    mi->parallel_mode);
+    table->field[48]->store(mode_name, strlen(mode_name), cs);
+  }
+  table->field[49]->store((uint32) mi->rli.get_sql_delay());
+  if (slave_sql_running_state == Relay_log_info::state_delaying_string)
+  {
+    time_t t= my_time(0), sql_delay_end= mi->rli.get_sql_delay_end();
+    table->field[50]->store((uint32)(t < sql_delay_end ? sql_delay_end - t : 0));
+  }
+  else
+    table->field[50]->store(STRING_WITH_LEN("NULL"), cs);
+
+  table->field[51]->store(slave_sql_running_state,
+                          strlen(slave_sql_running_state), cs);
+  table->field[52]->store((ulonglong) mi->total_ddl_groups, TRUE);
+  table->field[53]->store((ulonglong) mi->total_non_trans_groups, TRUE);
+  table->field[54]->store((ulonglong) mi->total_trans_groups, TRUE);
+  mysql_mutex_unlock(&mi->run_lock);
+  mysql_mutex_unlock(&mi->rli.err_lock);
+  mysql_mutex_unlock(&mi->err_lock);
+  mysql_mutex_unlock(&mi->rli.data_lock);
+  mysql_mutex_unlock(&mi->data_lock);
+  DBUG_RETURN(0);
+}
+#endif
+
+
+/*
+  Fill and store records into I_S.get_slave_status_record table
+
+  SYNOPSIS
+    get_slave_status_record()
+    thd                 thread handle
+    tables              table list struct(processed table)
+
+  RETURN
+    0   ok
+    #   error
+*/
+
+static int get_slave_status_record(THD *thd, TABLE_LIST *tables,
+                                   COND *cond __attribute__((unused)))
+{
+#ifndef HAVE_REPLICATION
+  my_ok(thd);
+  return false;
+#else
+  DBUG_ENTER("get_slave_status_record");
+  Master_info *mi;
+  uint i, elements;
+  /* Accept one of two privileges */
+  if (check_global_access(thd, PRIV_STMT_SHOW_SLAVE_STATUS))
+    DBUG_RETURN(1);
+
+  if (!master_info_index ||
+      !(elements= master_info_index->master_info_hash.records))
+    DBUG_RETURN(1);
+
+  mysql_mutex_lock(&LOCK_active_mi);
+  for (i= 0; i < elements; i++)
+  {
+    mi= (Master_info *) my_hash_element(&master_info_index->
+                                        master_info_hash, i);
+    mysql_mutex_lock(&mi->sleep_lock);
+    mi->users++;                                // Mark used
```

Comment:
If we are holding `LOCK_active_mi` for this whole loop, why play with the internal reference counting at all? Which is protected by `LOCK_active_mi`. We probably don't want to mess with it (so we'd neither `user++` nor `mi->release()`) as is the pattern of `show_all_master_info()`. Another inconsistency between this and `show all slaves status`, is the later sorts the results by name. Did you skip that intentionally?

Reply:
I agree, have removed. Well yes, I didn't want to sort, because we will have ability to use `WHERE` clause.